Mathlib/Data/Polynomial/RingDivision.lean

```diff
@@ -239,6 +239,19 @@ theorem isUnit_iff : IsUnit p ↔ ∃ r : R, IsUnit r ∧ C r = p :=
     fun ⟨_, hr, hrp⟩ => hrp ▸ isUnit_C.2 hr⟩
 #align polynomial.is_unit_iff Polynomial.isUnit_iff
 
+theorem not_isUnit_of_degree_pos [IsDomain R] (p : R[X])
```

Comment:
This is more generally true if `IsReduced R`, the proof should be very easy using [Polynomial.isUnit_iff_coeff_isUnit_isNilpotent](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Nilpotent.html#Polynomial.isUnit_iff_coeff_isUnit_isNilpotent) (but it requires more imports).

Reply:
Actually the generality seems incomparable: here we are in a `Semiring` that has `NoZeroDivisors` and `isUnit_iff_coeff_isUnit_isNilpotent` requires `CommRing`. Let me see if I can do a direct proof from `IsReduced`.

Reply:
@riccardobrasca I don't see how to relax the (commutative) ring condition for the `IsReduced` case, do you know?

Reply:
> @riccardobrasca I don't see how to relax the (commutative) ring condition for the `IsReduced` case, do you know?

Ah sorry, I automatically thought about `CommRing`. I guess we don't care about a super general result for `Semiring`, so let's just keep as it is.

Reply:
Even though one direction of [Polynomial.isUnit_iff_coeff_isUnit_isNilpotent](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Nilpotent.html#Polynomial.isUnit_iff_coeff_isUnit_isNilpotent) doesn't generalize to CommSemirings, [Polynomial.coeff_isUnit_isNilpotent_of_isUnit](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Nilpotent.html#Polynomial.coeff_isUnit_isNilpotent_of_isUnit) does generalize and is sufficient to prove `not_isUnit_of_degree_pos` for reduced rings. The original proof doesn't quite work: it still suffices to show that the coefficients lie in every prime ideal, but you can't necessarily take the quotient by an ideal in a semiring and then argue in the resulting domain.

My proof is inspired by [this answer](https://math.stackexchange.com/a/392604/12932), which shows if f is a unit and not a constant then its highest coefficient is nilpotent, and by induction on degree we just need to show f.eraseLead is also a unit. This follow from the general fact that if f+g is a unit and g is nilpotent in a CommSemiring, then f is a unit, with inverse given by $u=\sum_{n=0}^s\infty v^{n+1}g^n$, where v is the inverse of f+g; this is a finite sum because g is nilpotent. You can verify uf=1 by adding $v^m g^m=0$ to the end of $\sum_{n=0}^{m-1}v^{n+1}g^nf$ and keep cancelling $f+g$ with $v$ until only one term (1) is left.

Some fun exercises that seem to missing in mathlib include: if a+b=0 in a CommSemiring then a^2=b^2, and if moreover a^n=0, then b^n=0.

Reply:
Even though one direction of [Polynomial.isUnit_iff_coeff_isUnit_isNilpotent](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Nilpotent.html#Polynomial.isUnit_iff_coeff_isUnit_isNilpotent) doesn't generalize to CommSemirings, [Polynomial.coeff_isUnit_isNilpotent_of_isUnit](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Nilpotent.html#Polynomial.coeff_isUnit_isNilpotent_of_isUnit) does generalize and is sufficient to prove `not_isUnit_of_degree_pos` for reduced rings. The original proof doesn't quite work: it still suffices to show that the coefficients lie in every prime ideal, but you can't necessarily take the quotient by an ideal in a semiring and then argue in the resulting domain.

My proof is inspired by [this answer](https://math.stackexchange.com/a/392604/12932), which shows if f is a unit and not a constant then its highest coefficient is nilpotent, and by induction on degree we just need to show f.eraseLead is also a unit. This follow from the general fact that if f+g is a unit and g is nilpotent in a CommSemiring, then f is a unit, with inverse given by $u=\sum\nolimits_{n=0}^\infty v^{n+1}g^n$, where v is the inverse of f+g; this is a finite sum because g is nilpotent. You can verify uf=1 by adding $v^m g^m=0$ to the end of $\sum\nolimits_{n=0}^{m-1}v^{n+1}g^nf$ and keep cancelling $f+g$ with $v$ until only one term (1) is left.

Some fun exercises that seem to missing in mathlib include: if a+b=0 in a CommSemiring then a^2=b^2, and if moreover a^n=0, then b^n=0.

cc @XavierXarles since he's interested in semirings ...

Reply:
> Some fun exercises that seem to missing in mathlib include: if a+b=0 in a CommSemiring then a^2=b^2, and if moreover a^n=0, then b^n=0.

And this implies that if a+b=0 and b is nilpotent, then a is also nilpotent. More generally, if (a+b)^n=0 we can expand to get a^n+bx=0, and if b is nilpotent, then bx is nilpotent, so a^n, and therefore a, are nilpotent too. In summary, if (a+b) and b are nilpotent then so is a.

This can be used to generalize [Polynomial.isNilpotent_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Nilpotent.html#Polynomial.isNilpotent_iff) to CommSemirings: if f is nilpotent then clearly its leadingCoeff (equivalently, the leading term) is nilpotent. But f is the sum of f.eraseLead with the leading term, so f.eraseLead is nilpotent and we conclude by induction on degree.

Reply:
Nice ideas, thanks! So that this doesn't get lost, I copied your comments into #10539.